## Introductory steps & Data preprocessing

### Importing necessary libraries

In [105]:
import weaviate
from weaviate.classes.config import Configure, Property, DataType, VectorDistances

import os
import pandas as pd
import ast

### Loading the dataset

Loading dataset to a pandas dataframe. The dataset used for the workflow is the [LinkedIn Job Postings (2023 - 2024)](https://www.kaggle.com/datasets/arshkon/linkedin-job-postings) from Kaggle.

In [106]:
filename = "postings.csv"
df = pd.read_csv(filename)
df

job_id                     company_name  \
0           921716            Corcoran Sawyer Smith   
1          1829192                              NaN   
2         10998357           The National Exemplar    
3         23221523           Abrams Fensterman, LLP   
4         35982263                              NaN   
...            ...                              ...   
123844  3906267117                     Lozano Smith   
123845  3906267126                        Pinterest   
123846  3906267131                     EPS Learning   
123847  3906267195  Trelleborg Applied Technologies   
123848  3906267224                        Solugenix   

                                                    title  \
0                                   Marketing Coordinator   
1                       Mental Health Therapist/Counselor   
2                             Assitant Restaurant Manager   
3       Senior Elder Law / Trusts and Estates Associat...   
4                                      Service Technician   
...                                                   ...   
123844                   Title IX/Investigations Attorney   
123845       Staff Software Engineer, ML Serving Platform   
123846               Account Executive, Oregon/Washington   
123847                       Business Development Manager   
123848                  Marketing Social Media Specialist   

                                              description  max_salary  \
0       Job descriptionA leading real estate firm in N...        20.0   
1       At Aspen Therapy and Wellness , we are committ...        50.0   
2       The National Exemplar is accepting application...     65000.0   
3       Senior Associate Attorney - Elder Law / Trusts...    175000.0   
4       Looking for HVAC service tech with experience ...     80000.0   
...                                                   ...         ...   
123844  Our Walnut Creek office is currently seeking a...    195000.0   
123845  About Pinterest:\n\nMillions of people across ...         NaN   
123846  Company Overview\n\nEPS Learning is a leading ...         NaN   
123847  The Business Development Manager is a 'hunter'...         NaN   
123848  Marketing Social Media Specialist - $70k – $75...     75000.0   

       pay_period                 location  company_id  views  med_salary  \
0          HOURLY            Princeton, NJ   2774458.0   20.0         NaN   
1          HOURLY         Fort Collins, CO         NaN    1.0         NaN   
2          YEARLY           Cincinnati, OH  64896719.0    8.0         NaN   
3          YEARLY        New Hyde Park, NY    766262.0   16.0         NaN   
4          YEARLY           Burlington, IA         NaN    3.0         NaN   
...           ...                      ...         ...    ...         ...   
123844     YEARLY         Walnut Creek, CA     56120.0    1.0         NaN   
123845        NaN            United States   1124131.0    3.0         NaN   
123846        NaN              Spokane, WA  90552133.0    3.0         NaN   
123847        NaN     Texas, United States   2793699.0    4.0         NaN   
123848     YEARLY  San Juan Capistrano, CA     43325.0    2.0         NaN   

        ...        expiry closed_time  formatted_experience_level  \
0       ...  1.715990e+12         NaN                         NaN   
1       ...  1.715450e+12         NaN                         NaN   
2       ...  1.715870e+12         NaN                         NaN   
3       ...  1.715488e+12         NaN                         NaN   
4       ...  1.716044e+12         NaN                         NaN   
...     ...           ...         ...                         ...   
123844  ...  1.716163e+12         NaN            Mid-Senior level   
123845  ...  1.716164e+12         NaN            Mid-Senior level   
123846  ...  1.716164e+12         NaN            Mid-Senior level   
123847  ...  1.716165e+12         NaN                         NaN   
123848  ...  1.716165e+12         NaN            Mid-Seni

### Preprocessing the dataset

In our dataset, we merge the columns company_name, title, skills_desc and description in a new column named merged_col. 
Before every company name in the merged_col we add the prefix "Company Name:". Similarly, before the titles we add the prefix "Job Title:" and before the skills_desc and description we add the "Job Description:" prefix.

In [107]:
df['company_name'] = df['company_name'].fillna('')
df['title'] = df['title'].fillna('')
df['skills_desc'] = df['skills_desc'].fillna('')
df['description'] = df['description'].fillna('')
df['job_posting_url'] = df['job_posting_url'].fillna('')

df["merged_col"] = "Company Name: " + df["company_name"] + "\nJob Title: " + df["title"] + "\nJob Description:\n" + df["skills_desc"] + "\n" + df["description"]

df1 = df[["company_name", "title", "description", "skills_desc", "job_posting_url", "merged_col"]]
df1

company_name  \
0                 Corcoran Sawyer Smith   
1                                         
2                The National Exemplar    
3                Abrams Fensterman, LLP   
4                                         
...                                 ...   
123844                     Lozano Smith   
123845                        Pinterest   
123846                     EPS Learning   
123847  Trelleborg Applied Technologies   
123848                        Solugenix   

                                                    title  \
0                                   Marketing Coordinator   
1                       Mental Health Therapist/Counselor   
2                             Assitant Restaurant Manager   
3       Senior Elder Law / Trusts and Estates Associat...   
4                                      Service Technician   
...                                                   ...   
123844                   Title IX/Investigations Attorney   
123845       Staff Software Engineer, ML Serving Platform   
123846               Account Executive, Oregon/Washington   
123847                       Business Development Manager   
123848                  Marketing Social Media Specialist   

                                              description  \
0       Job descriptionA leading real estate firm in N...   
1       At Aspen Therapy and Wellness , we are committ...   
2       The National Exemplar is accepting application...   
3       Senior Associate Attorney - Elder Law / Trusts...   
4       Looking for HVAC service tech with experience ...   
...                                                   ...   
123844  Our Walnut Creek office is currently seeking a...   
123845  About Pinterest:\n\nMillions of people across ...   
123846  Company Overview\n\nEPS Learning is a leading ...   
123847  The Business Development Manager is a 'hunter'...   
123848  Marketing Social Media Specialist - $70k – $75...   

                                              skills_desc  \
0       Requirements: \n\nWe are seeking a College or ...   
1                                                           
2       We are currently accepting resumes for FOH - A...   
3       This position requires a baseline understandin...   
4                                                           
...                                                   ...   
123844                                                      
123845                                                      
123846                                                      
123847                                                      
123848                                                      

                                          job_posting_url  \
0       https://www.linkedin.com/jobs/view/921716/?trk...   
1       https://www.linkedin.com/jobs/view/1829192/?tr...   
2       https://www.linkedin.com/jobs/view/10998357/?t...   
3       https://www.linkedin.com/jobs/view/23221523/?t...   
4       https://www.linkedin.com/jobs/view/35982263/?t...   
...                                                   ...   
123844  https://www.linkedin.com/jobs/view/3906267117/...   
123845  https://www.linkedin.com/jobs/view/3906267126/...   
123846  https://www.linkedin.com/jobs/view/3906267131/...   
123847  https://www.linkedin.com/jobs/view/3906267195/...   
123848  https://www.linkedin.com/jobs/view/3906267224/...   

                                               merged_col  
0       Company Name: Corcoran Sawyer Smith\nJob Title...  
1       Company Name: \nJob Title: Mental Health Thera...  
2       Company Name: The National Exemplar \nJob Titl...  
3       Company Name: Abrams Fensterman, LLP\nJob Titl...  
4       Company Name: \nJob Title:  Service Technician...  
...                                                   ...  
123844  Company Name: Lozano Smith\nJob Title: Title I...  
123845  Company Name: Pinterest\nJob Title: Staff Soft...  
123846  Company Name: EPS Learning\nJo

We same the new version of our dataframe to a new csv with the name "postings_new.csv".

In [4]:
df1.to_csv("postings_new.csv", index=False)

## Collection creation and data import

### Connect to Weaviate Cloud with an API

Before running the code below, you need to add OPENAI_API_KEY, WCD_URL and WCD_API_KEY as an environmental variable.

In [ ]:
# If running the code in a virtual environment, you need to uncomment the code below and add your OPENAI_API_KEY. Else, you can set it from terminal.
# os.environ["OPENAI_API_KEY"] = "YOUR_API_KEY"
# os.environ["WCD_URL"] = "YOUR_WCD_URL"
# os.environ["WCD_API_KEY"] = "YOUR_WCD_API_KEY"


In [3]:
client = weaviate.connect_to_wcs(
    cluster_url=os.getenv("WCD_URL"),
    auth_credentials=weaviate.auth.AuthApiKey(os.getenv("WCD_API_KEY")),
    headers={
        "X-OpenAI-Api-Key": os.environ["OPENAI_API_KEY"]  # Replace with your inference API key
    }
)

### Collection creation

Creating the collection with name "LinkedIn_Job_Postings". It has 6 properties, but we will only vectorize the "merged_col" property using the "text-embedding-3-small" vectorizer from OpenAI.

In [9]:
client.collections.create(
    "LinkedIn_Job_Postings",
    vectorizer_config=Configure.Vectorizer.text2vec_openai(model="text-embedding-3-small"),
    vector_index_config=Configure.VectorIndex.hnsw(distance_metric=VectorDistances.COSINE),
    properties=[
        Property(name="company_name", data_type=DataType.TEXT,skip_vectorization=True),
        Property(name="title", data_type=DataType.TEXT,skip_vectorization=True),
        Property(name="description", data_type=DataType.TEXT, skip_vectorization=True),
        Property(name="skills_desc", data_type=DataType.TEXT, skip_vectorization=True),
        Property(name="merged_col", data_type=DataType.TEXT, skip_vectorization=False),
        Property(name="job_posting_url", data_type=DataType.TEXT, skip_vectorization=True),
    ],
    
)

### Import data

Let's populate our database with data. We use the data from the "postings_new.csv" that we created before and add them in batches.

In [10]:
collection = client.collections.get("LinkedIn_Job_Postings")

with client.batch.fixed_size(batch_size=200) as batch:
    with pd.read_csv(
        "postings_new.csv",
        usecols=["company_name", "title", "description", "skills_desc", "job_posting_url", "merged_col"],
        chunksize=100,
    ) as csv_iterator:
        for chunk in csv_iterator:
            for index, row in chunk.iterrows():
                batch.add_object(
                    collection="LinkedIn_Job_Postings",
                        properties = {
                            "company_name": row["company_name"],
                            "title": row["title"],
                            "description": row["description"],
                            "skills_desc": row["skills_desc"],
                            "job_posting_url": row["job_posting_url"],
                            "merged_col": row["merged_col"]
                         }
                )

## Finding the job openings that are related to machine learning or AI

Let's get the collection and search for the jobs realted to ML or AI.

In [41]:
collection = client.collections.get("LinkedIn_Job_Postings")

We create a dataframe where we will collect the unique job offers based on all of our different methods of search.

In [100]:
df_ml = pd.DataFrame(columns=('uuid', 'company_name', 'title', 'job_posting_url'))

These are the terms that we are going to search for: "Machine Learning AI", "Machine Learning", "Data Science", "Artificial Intelligence". We will skip searching for abbreviations as they may return irrelevant jobs.

In [101]:
queries = ["Machine Learning AI", "Machine Learning", "Data Science", "Artificial Intelligence"]

### Vector search

In [102]:
for q in queries:
    response = collection.query.near_text(query=q)
    for o in response.objects:
        new_row = {"uuid": o.uuid, "company_name": o.properties["company_name"], "title": o.properties["title"], "job_posting_url": o.properties["job_posting_url"]}
        if o.uuid not in df_ml["uuid"].values:
            # Create a DataFrame from the new row
            new_row_df = pd.DataFrame([new_row])
            
            # Concatenate the new row to the existing DataFrame
            df_ml = pd.concat([df_ml, new_row_df], ignore_index=True)
df_ml[["company_name", "title", "job_posting_url"]]


company_name  \
0                     hackajob   
1                     hackajob   
2                     hackajob   
3                     hackajob   
4          McCulloh Consulting   
..                         ...   
263            Husch Blackwell   
264                   Deloitte   
265  Hyperion Technologies LLC   
266                     Photon   
267                   Deloitte   

                                                 title  \
0                            Machine Learning Engineer   
1                     Artificial Intelligence Engineer   
2                                       Data Scientist   
3                                       Data Scientist   
4                            Machine Learning Engineer   
..                                                 ...   
263  Legal Artificial Intelligence Implementation A...   
264                              AI Solution Architect   
265                              AI Engineer/Architect   
266                                   Lead AI Engineer   
267  Data Scientist - GenAI - Solution Specialist -...   

                                       job_posting_url  
0    https://www.linkedin.com/jobs/view/3905881936/...  
1    https://www.linkedin.com/jobs/view/3905888074/...  
2    https://www.linkedin.com/jobs/view/3905875688/...  
3    https://www.linkedin.com/jobs/view/3905875652/...  
4    https://www.linkedin.com/jobs/view/3903865758/...  
..                                                 ...  
263  https://www.linkedin.com/jobs/view/3904988939/...  
264  https://www.linkedin.com/jobs/view/3884830087/...  
265  https://www.linkedin.com/jobs/view/3903445676/...  
266  https://www.linkedin.com/jobs/view/3886883337/...  
267  https://www.linkedin.com/jobs/view/3899532438/...  

[268 rows x 3 columns]

So based on our vector search we have 268 job postings related to ML and AI.

### Keyword search

Let's query the terms "Machine Learning" and "AI" using the BM25 (Keyword) search and return the company title, job title and the url of the job offer. Weaviate will look for objects that contain our search terms in the property "merged_col" and will rank the results based on how many times the search terms appear in the "merged_col" property of the objects.

In [103]:
for q in queries:
    response = collection.query.bm25(
    query=q,
    query_properties=["merged_col"]
)
    for o in response.objects:
        new_row = {"uuid": o.uuid, "company_name": o.properties["company_name"], "title": o.properties["title"], "job_posting_url": o.properties["job_posting_url"]}
        if o.uuid not in df_ml["uuid"].values:
            # Create a DataFrame from the new row
            new_row_df = pd.DataFrame([new_row])
            
            # Concatenate the new row to the existing DataFrame
            df_ml = pd.concat([df_ml, new_row_df], ignore_index=True)
df_ml[["company_name", "title", "job_posting_url"]]

company_name                                              title  \
0               hackajob                          Machine Learning Engineer   
1               hackajob                   Artificial Intelligence Engineer   
2               hackajob                                     Data Scientist   
3               hackajob                                     Data Scientist   
4    McCulloh Consulting                          Machine Learning Engineer   
..                   ...                                                ...   
491            Microsoft                   Senior Technical Program Manager   
492            Greystone                       Digital Automation Architect   
493              Cargill         Manager, Global HR Systems & Process Owner   
494             Polygraf            AI & ML Engineer – Audio Specialization   
495           RedBalloon  Director of IT Architecture (with R&D Focus)- ...   

                                       job_posting_url  
0    https://www.linkedin.com/jobs/view/3905881936/...  
1    https://www.linkedin.com/jobs/view/3905888074/...  
2    https://www.linkedin.com/jobs/view/3905875688/...  
3    https://www.linkedin.com/jobs/view/3905875652/...  
4    https://www.linkedin.com/jobs/view/3903865758/...  
..                                                 ...  
491  https://www.linkedin.com/jobs/view/3901355763/...  
492  https://www.linkedin.com/jobs/view/3904382623/...  
493  https://www.linkedin.com/jobs/view/3904993645/...  
494  https://www.linkedin.com/jobs/view/3903808878/...  
495  https://www.linkedin.com/jobs/view/3902741272/...  

[496 rows x 3 columns]

Keyword search added 228 extra jobs postings related to ML and AI. So, now we have 496. Let's save them to a csv.

In [104]:
df_ml.to_csv('ml_jobs.csv', index=False)

### Closing the client

After ending with our experiments we can close the client

In [ ]:
client.close() 